src/node_api.cc

```diff
@@ -152,10 +153,22 @@ class BufferFinalizer : private Finalizer {
 };
 
 inline napi_env NewEnv(v8::Local<v8::Context> context,
-                       const std::string& module_filename) {
+                       const std::string& module_filename,
+                       int32_t module_api_version) {
   node_napi_env result;
 
-  result = new node_napi_env__(context, module_filename);
+  // Validate module_api_version.
+  if (module_api_version < NODE_API_DEFAULT_MODULE_API_VERSION) {
+    module_api_version = NODE_API_DEFAULT_MODULE_API_VERSION;
+  } else if (module_api_version > NAPI_VERSION &&
+             module_api_version != NAPI_VERSION_EXPERIMENTAL) {
+    node::Environment* node_env = node::Environment::GetCurrent(context);
+    CHECK_NOT_NULL(node_env);
+    node_env->ThrowError("Node API module targets unsupported version.");
```

Comment:
Can we make this more verbose?

```suggestion
    node_env->ThrowError(
        module_filename + "requires Node-API version " +
        std::to_string(module_api_version) + ", but this version of Node.js "
        "only supports version " #NAPI_VERSION " add-ons.");
```

Reply:
Good suggestion! I applied your code and extracted it into a separate method since we need it in two different places.